src/stan/callbacks/json_writer.hpp

```diff
@@ -0,0 +1,401 @@
+   * @return The processed string.
+   */
+  std::string process_string(const std::string& value) {
+    static constexpr std::array<char, 11> chars_to_escape
+        = {'\\', '"', '/', '\b', '\f', '\n', '\r', '\t', '\v', '\a', '\0'};
+    static constexpr std::array<const char*, 11> chars_to_replace
+        = {"\\\\", "\\\"", "\\/", "\\b", "\\f", "\\n",
+           "\\r",  "\\t",  "\\v", "\\a", "\\0"};
+    // Replacing every value leads to 2x the size
+    std::string new_value(value.size() * 2, 'x');
+    std::size_t pos = 0;
+    std::size_t count = 0;
+    std::size_t prev_pos = 0;
+    while ((pos = value.find_first_of(chars_to_escape.data(), pos, 10))
+           != std::string::npos) {
+      for (int i = prev_pos; i < pos; ++i) {
+        new_value[i + count] = value[i];
+      }
+      int idx
+          = strchr(chars_to_escape.data(), value[pos]) - chars_to_escape.data();
+      new_value[pos + count] = chars_to_replace[idx][0];
+      new_value[pos + count + 1] = chars_to_replace[idx][1];
+      pos += 1;
+      count++;
+      prev_pos = pos;
+    }
+    for (int i = prev_pos; i < value.size(); ++i) {
+      new_value[i + count] = value[i];
+    }
+    // Shrink any unused space
+    new_value.resize(value.size() + count);
+    return new_value;
+  }
+
+  /**
+   * Writes key plus colon for key-value pair.
+   *
+   * @param[in] key member name.
+   */
+  void write_key(const std::string& key) { *output_ << "\"" << key << "\" : "; }
+  /**
+   * Writes a set of comma separated strings.
+   * Strings are cleaned to escape special characters.
+   *
+   * @param[in] v Values in a std::vector
+   */
+  void write_vector(const std::vector<std::string>& v) {
+    if (v.empty()) {
+      return;
+    }
+    *output_ << "[ ";
+    auto last = v.end();
+    --last;
+    for (auto it = v.begin(); it != last; ++it) {
+      *output_ << process_string(*it) << ", ";
+    }
+    *output_ << v.back() << " ]";
+  }
+
+  /**
+   * Writes a set of comma separated double values.
+   *
+   * @param[in] v Values in a std::vector
+   */
+  void write_vector(const std::vector<double>& v) {
+    if (v.empty()) {
+      return;
+    }
+    *output_ << "[ ";
+    auto last = v.end();
+    --last;
+    for (auto it = v.begin(); it != last; ++it) {
+      write_element(*it);
+      *output_ << ", ";
+    }
+    write_element(v.back());
+    *output_ << " ]";
+  }
+
+  /**
+   * Writes a set of comma separated values.
+   *
+   * @param[in] v Values in a std::vector
+   */
+  template <class T>
+  void write_vector(const std::vector<T>& v) {
+    if (v.empty()) {
+      return;
+    }
+    *output_ << "[ ";
+    auto last = v.end();
+    --last;
+    for (auto it = v.begin(); it != last; ++it) {
+      *output_ << *it << ", ";
+    }
+    *output_ << v.back() << " ]";
+  }
+
+  /**
+   * Writes a single value.  Corrects capitalization for inf and nans.
+   *
+   * @param[in] v value
+   */
+  void write_element(double v) {
```

Comment:
`unlikely(...)` is since C++20 - what is allowed in Stan these days?

Reply:
We're only on 14 still, but we have a macro which does the same thing (and is also called `unlikely`) when it can: https://github.com/stan-dev/math/blob/develop/stan/math/prim/meta/compiler_attributes.hpp

We use it e.g. here:
https://github.com/stan-dev/stan/blob/1b126f5c8da9d374d0bee934779a6600112ed08c/src/stan/io/deserializer.hpp#L179

Reply:
using this everywhere - for double, Eigen, and complex values.